### PR TITLE
Make TOC easier to maintain.

### PR DESCRIPTION
**Pull request by levino**
- Make the TOC truely dynamic by changing data model and graphQL query
- Refactor stateful and difficult to read TOC component

The functionality of the boilerplate should not be affected by this. Also everything should still look the same from the user perspective.

### CODE REVIEW

File: content/lessons/table_of_contents.json
Comment (levino) [inline content/lessons/table_of_contents.json:3]: I did not see where this would be used. Removed it.

File: content/lessons/table_of_contents.json
Comment (levino) [inline content/lessons/table_of_contents.json:5]: These "named entries" make no sense because we have to repeat the whole datastructure with these names in the GraphQL query for the TOC. This should be an array of arbitrary length that is evaluated dynamically and not expected to have hardcoded keys in it.

File: content/lessons/table_of_contents.json
Comment (levino) [inline content/lessons/table_of_contents.json:6]: The very same point as with `"one"`

File: content/lessons/table_of_contents.json
Comment (levino) [inline content/lessons/table_of_contents.json]: This is not a blog, but a documentation. I think `"post"` does not make sense. I propose "`entry`". Happy to discuss naming here.

File: content/lessons/table_of_contents.json
Comment (levino) [inline content/lessons/table_of_contents.json]: Make the title a proper string and do not parse a JSON key to get the String to be displayed in the HTML. The approach chosen before (parsing the json key - not value - via [this function](https://github.com/ericwindmill/gatsby-starter-docs/blob/1967f125f938bb25b72ea00f6dcc5c5eec7d6520/src/components/Layout/TableOfContents.jsx#L13-L15)) is a quite bad idea. How would you ever want to account for chapter titles like `"The BigNum Problem"` or i18n?

File: content/lessons/table_of_contents.json
Comment (levino) [inline content/lessons/table_of_contents.json]: Same here as for the chapter title.s

File: src/components/Layout/TableOfContents.jsx
Comment (levino) [inline src/components/Layout/TableOfContents.jsx]: This method should **not** be in the scope of the class. If you need such a function, it should be declared outside of the class at the top of the document like so:

```js
import React from "react"
import Link from 'gatsby-link'
import styled from 'styled-components'

const formatChapterTitle = (title) => title.split('_').map(word => word[0].toUpperCase() + word.substring(1)).join(' ')

class TableOfContents extends React.Component {
  constructor() {
    super();
    this.nodeListItemsToRender = [];
    this.currentLevel = 0;
  }

  addSubchapterJSX(node) {
    this.nodeListItemsToRender.push(
      <SubchapterLIContainer key={node}>
        <h5>
          {formatChapterTitle(node)}
        </h5>
      </SubchapterLIContainer>
    );
  }

  addChapterJSX(node) {
    this.nodeListItemsToRender.push(
      <ChapterLIContainer key={node}>
        <h5>
          {formatChapterTitle(node)}
        </h5>
      </ChapterLIContainer>
    )
  }
...
```
You are **not** using `this` in the function. So then do not clutter the classes interface with it. In theory it is also a performance issue because the member method would be created for each instance. In practice V8 will optimize this away, but relying on this is bad practice.

As I propose a better approach for handling the chapter titles, I was able to remove the function altogether.

File: src/components/Layout/TableOfContents.jsx
Comment (levino) [inline src/components/Layout/TableOfContents.jsx]: Only use classes when you need state. You do not need state here or in Gatsby in general. So use stateless (aka functional) components. It makes everything so much easier.

File: src/components/Layout/TableOfContents.jsx
Comment (levino) [inline src/components/Layout/TableOfContents.jsx]: State mutation is difficult to read, to debug and to understand. As you can see the same result can be achieved with just evaluating one (big and composed) function without ever mutating any variable.

File: src/components/Layout/TableOfContents.jsx
Comment (levino) [inline src/components/Layout/TableOfContents.jsx:71]: I admit that my approach does not (in this simple version) allow for arbitrary depth of the Chapter - Subchapter - ... construct. But I also think that it does not make sense to support this in the boilerplate. Also it is easy to adjust my approach with a little recursion to allow for any kind of nesting some sick mind could come up with.

File: src/components/Layout/TableOfContents.jsx
Comment (levino) [inline src/components/Layout/TableOfContents.jsx]: The **order** entries in the menu for the TOC **matters**, I would say. So this should an **ordered list**. I also think that  using a "flat" list (chapter titles, subchapter titles, entry titles are all each just on `li` in the top level list) is a bad approach. I created sublists for the individual entries to have a more finegrained logic. This will be very helpful when someone wants to do more complex styling (some padding around the `EntryList`s for example).

File: src/templates/lesson.jsx
Comment (levino) [inline src/templates/lesson.jsx]: We are not interested in anything else but the chapters anyhow so we can already filter for them here.

File: src/templates/lesson.jsx
Comment (levino) [inline src/templates/lesson.jsx]: The query is now very general and allows for any number of chapters and subchapters in the `table_of_contens.json`. Just one place where you have to add a new chapter, not two which have to be in sync (which was driving me crazy).

**Review by levino — COMMENTED** [review]
I left some additional remarks in the code to explain my thinking.

File: src/components/Layout/TableOfContents.jsx
Comment (ericwindmill) [reply src/components/Layout/TableOfContents.jsx:71]: @Levino I agree with every comment in here, except this one. My reasoning for redoing the ToC in the way that it is now is because I needed the freedom to change the order and nesting of ToC without much work.

I agree that repeating the structure in the GraphQL is an issue. It's getting cumbersome as my site grows (the one I built this for). It also leaves a lot of room for silly typo errors.

Can we think of a way that allows for an arbitrary number of nested levels, but maintain all the benefits that your changes would provide?

Comment (levino) [reply src/components/Layout/TableOfContents.jsx:71]: Yes we potentially can. I will think about this. But even then the code will be more easy to read with recursion. Will take a couple of days for me to come back to this though.

**Comment by levino** [comment]
This is work in progress. Do not merge.

**Comment by levino** [comment]
Your concerns should have been adressed. Please review. Also note: This contains changes from https://github.com/Levino/gatsby-starter-docs/tree/feature/16-nicer-code-styles

**Comment by levino** [comment]
Also I added a configuration to run linting in CI (circleci). It is for free. Once you merge this branch in, you can run builds on circleci and have the status here as requirement for merging PRs. I think this is a good idea because the code failed the linting and I had to fix a lot of stuff I never touched.

**Comment by levino** [comment]
Because of the prettification the diff is huge. I recommend merging #23 first. I will then update to minimize the diff.

File: src/templates/lesson.jsx
Comment (levino) [inline src/templates/lesson.jsx]: Actually I wanted to make this even nicer, but it is not possible.

First read [this](https://github.com/facebook/graphql/issues/91): Arbitrary recursion is not possible within a graphql query.

I tried to make this more DRY like so:

```graphql
{
  tableOfContents: lessonsJson {
     	...ChaptersRecursive
	}
}

fragment ChaptersRecursive on LessonsJson {
  chapters {
    ...chapterFields
    chapters {
      ...chapterFields
      chapters {
        ...chapterFields
      }
    }
  }
}

fragment chapterFields on Chapters {
  title
  entries {
    entry {
      id
      childMarkdownRemark {
        fields {
          slug
        }
        frontmatter {
          title
        }
      }
    }
  }
}
```

But the st***d 'gatsby-transformer-json' creates a type for each individual `chapter` ocurrence instead of one `Chapter` type. You can paste the above in your local graphiql page to see what I mean. One would have to manually set the type of the chapters to a type `Chapter`, but I do not understand how to configure 'gatsby-transformer-json'.

File: src/pages/about.jsx
Comment (levino) [inline src/pages/about.jsx:2]: This stuff was unused so linter complained.

**Comment by levino** [comment]
Readding the comment here because it is now considered "outdated":

Actually I wanted to make the graphql query even nicer, but it is not possible.

First read [this](https://github.com/facebook/graphql/issues/91): Arbitrary recursion is not possible within a graphql query.

I tried to make this more DRY like so:

```graphql
{
  tableOfContents: lessonsJson {
     	...ChaptersRecursive
	}
}

fragment ChaptersRecursive on LessonsJson {
  chapters {
    ...chapterFields
    chapters {
      ...chapterFields
      chapters {
        ...chapterFields
      }
    }
  }
}

fragment chapterFields on Chapters {
  title
  entries {
    entry {
      id
      childMarkdownRemark {
        fields {
          slug
        }
        frontmatter {
          title
        }
      }
    }
  }
}
```

But the st***d 'gatsby-transformer-json' creates a type for each individual `chapter` ocurrence instead of one `Chapter` type. You can paste the above in your local graphiql page to see what I mean. One would have to manually set the type of the chapters to a type `Chapter`, but I do not understand how to configure 'gatsby-transformer-json'.

**Comment by levino** [comment]
Good to go in, imo.

**Review by ericwindmill — APPROVED** [review]

**Comment by ericwindmill** [comment]
Sincere thanks my man. It's a huge approvement. I think future users will be very happy with this.

**Pull request merged**
